tools/devservices_healthcheck.py

```diff
@@ -1,47 +1,183 @@
+    id: str
+    container_name: str
+    check_by_default: bool
+    check: Callable[[], None]
+    deps: list[str]
+    retries: int
+    timeout_secs: int
+
+    def __init__(
+        self, id, container_name, check_by_default, check=None, deps=None, retries=3, timeout_secs=5
+    ):
+        self.id = id
+        self.container_name = container_name
+        self.check_by_default = check_by_default
+        self.check = check
+        self.deps = deps or []
+        self.retries = retries
+        self.timeout_secs = timeout_secs
+
+    def check_container(self):
+        response = subprocess.run(
+            ("docker", "container", "inspect", "-f", "'{{.State.Status}}'", self.container_name),
+            capture_output=True,
+            text=True,
+        )
+        if response.stdout.strip() != "'running'":
+            raise HealthcheckError(f"Container '{self.container_name}' is not running.")
+
+
+def check_kafka():
+    # sentry_zookeeper:2181 doesn't work in CI, but 127.0.0.1 doesn't work locally
+    zookeeper_origin = "sentry_zookeeper"
+    if os.getenv("CI") == "true":
+        zookeeper_origin = "127.0.0.1"
+
+    subprocess.run(
+        (
+            "docker",
+            "exec",
+            "sentry_kafka",
+            "kafka-topics",
+            "--zookeeper",
+            f"{zookeeper_origin}:2181",
+            "--list",
+        ),
+        check=True,
+    )
+
+
+def check_postgres():
+    subprocess.run(
+        ("docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"), check=True
+    )
+
+
+# Available health checks
+all_service_healthchecks = {
+    "postgres": HealthCheck(
+        "postgres",
+        "sentry_postgres",
+        True,
+        lambda: subprocess.run(
+            ["docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"], check=True
+        ),
+    ),
+    "kafka": HealthCheck(
+        "kafka",
+        "sentry_kafka",
+        os.getenv("NEED_KAFKA") == "true",
+        check_kafka,
+        deps=["zookeeper"],
+    ),
+    "zookeeper": HealthCheck(
+        "zookeeper",
+        "sentry_zookeeper",
+        os.getenv("NEED_KAFKA") == "true",
+    ),
+}
 
-        if returncode != 0:
-            time.sleep(timeout)
+
+def run_with_retries(cmd: Callable[[], object], retries: int, timeout: int) -> None:
+    for retry in range(1, retries + 1):
+        try:
+            cmd()
+        except Exception as e:
+            if retry == retries:
+                print(f"Command failed, no more retries: {e}")
+                raise HealthcheckError(f"Command failed: {e}")
+            else:
+                print(f"Command failed, retrying in {timeout}s (attempt {retry+1} of {retries})...")
+                time.sleep(timeout)
         else:
             return
 
-    raise SystemExit(1)
 
+def get_services_to_check(id: str) -> list[str]:
+    if all_service_healthchecks.get(id) is None:
+        raise HealthcheckError(f"Service '{id}' does not have a health check")
```

Comment:
I would maybe just let the `KeyError` raise -- it's a _programmer_ error at that point and not a _user_ error

Reply:
This is intended to catch human errors: `python3 -m tools.devservices_healthcheck --service="snabu"`

Reply:
do we expect users to run this directly?

Reply:
cool now we can remove this with the `choices` setup below \o/